Firmware/pat9125.c

```diff
@@ -234,39 +254,48 @@ uint8_t pat9125_update_bs(void)
 uint8_t pat9125_rd_reg(uint8_t addr)
 {
 	uint8_t data = 0;
-#ifdef PAT9125_SWSPI
+#if defined(PAT9125_SWSPI)
 	swspi_start();
 	swspi_tx(addr & 0x7f);
 	data = swspi_rx();
 	swspi_stop();
-#endif //PAT9125_SWSPI
-#ifdef PAT9125_SWI2C
+#elif defined(PAT9125_SWI2C)
 	if (!swi2c_readByte_A8(PAT9125_I2C_ADDR, addr, &data)) //NO ACK error
-	{
-		pat9125_PID1 = 0xff;
-		pat9125_PID2 = 0xff;
-		return 0;
-	}
-#endif //PAT9125_SWI2C
+        goto error;
+#elif defined(PAT9125_I2C)
+	if (twi_rw8(PAT9125_I2C_ADDR,TW_WRITE,&addr) ||
+        twi_rw8(PAT9125_I2C_ADDR,TW_READ,&data))
+        goto error;
+#endif
 	return data;
+
+ error:
+    pat9125_PID1 = 0xff;
+    pat9125_PID2 = 0xff;
+    return 0;
 }
 
 void pat9125_wr_reg(uint8_t addr, uint8_t data)
 {
-#ifdef PAT9125_SWSPI
+#if defined(PAT9125_SWSPI)
 	swspi_start();
 	swspi_tx(addr | 0x80);
 	swspi_tx(data);
 	swspi_stop();
-#endif //PAT9125_SWSPI
-#ifdef PAT9125_SWI2C
+#elif defined(PAT9125_SWI2C)
 	if (!swi2c_writeByte_A8(PAT9125_I2C_ADDR, addr, &data)) //NO ACK error
-	{
-		pat9125_PID1 = 0xff;
-		pat9125_PID2 = 0xff;
-		return;
-	}
-#endif //PAT9125_SWI2C
+        goto error;
+#elif defined(PAT9125_I2C)
+	if (twi_rw8(PAT9125_I2C_ADDR,TW_WRITE,&addr) ||
+        twi_rw8(PAT9125_I2C_ADDR,TW_READ,&data))
```

Comment:
This isn't right - it doesn't actually write data to the device; it sets the address and then does a read on the same address - as evidenced by the printer startup:

```
PAT9125_RES_X=20. <-- should be 0
PAT9125_RES_Y=20. <-- should be 240. 
PAT9125_init:1.
```

Moreover, I don't think this is necessarily I2C compliant. You need to send the data payload in the same start-stop frame as the address byte. Look at swi2c_readByte_A8 to see this.

Reply:
I probably shaved this while trying to save space, but you're definitely right.
I'll test this tomorrow on a real MK3.

Reply:
Should be able to just modify twi_rw8 to take an additional pointer parameter to the byte to send or be read into. Would save the overhead of two successive function calls there and some i2c delays.

Reply:
*Should* be correct now. I've split the two functions back, which I think it's more readable anyway.

Reply:
Yup, all seems to be as expected now!

```
<snip PID/VID check>
X_IN:00ea0100
START
TX_IN:0dea1000
ADDRESS 13
TX_IN:00ea1000
Set 13 to 0
TX_IN:01ea0200
STOP
TX_IN:00ea0100
START
TX_IN:0eea1000
ADDRESS 14
TX_IN:f0ea1000
Set 14 to 240
TX_IN:01ea0200
STOP
TX_IN:00ea0100
START
TX_IN:0dea1000
ADDRESS 13
TX_IN:00eb0100
START
TX_IN:ebeb2000
READ 13
TX_IN:01eb0200
STOP
PAT9125_RES_X=0.
TX_IN:00ea0100
START
TX_IN:0eea1000
ADDRESS 14
TX_IN:00eb0100
START
TX_IN:ebeb2000
READ 14
TX_IN:01eb0200
STOP
PAT9125_RES_Y=240.
PAT9125_init:1.
```